# INSTRUCTIONS
1. Fork this repository into your GitHub folder
2. Clone YOUR COPY into your laptop
3. cd into the cloned folder
4. start jupyter notebook
5. Open the appropriate exam (Exam Week 1, or Exam Week 2)
6. Create a new Jupyter Notebook using the Bash kernel (for week 1 exam) or python kernel (week 2 exam)
7. In this new notebook, create a documentation ("Markup") box, and a code box, for each question
    *use the documentation box to explain the logic of your answer
    *the code box to show the code of your answer
8. when you have finished answering all problems, save the notebook
9. push the notebook to GitHub
10. email me your GitHub username so that I can retrieve your answers and grade them.
11. go for a beer!

# BACKGROUND INFORMATION

I have retrieved the Locus Germplasm Phenotype file from The Arabidiosis Information Resource (TAIR - one of the primary _Arabidopsis_ databases globally) (https://www.arabidopsis.org/download/index-auto.jsp?dir=/download_files/Polymorphisms_and_Phenotypes ).  The file contains information about the phenotypes associated with certain _Arabodopsis_ mutations.

The file is in the current folder - called [Locus_Germplasm_Phenotype_20130122.txt](./Locus_Germplasm_Phenotype_20130122.txt)

The file format is "tab-separated values".  This means that each data field is separated by a "tab" character.  **There is one line at the beginning of the file -the "header" - that describes the type of data in each column.**  

The first column is the "Arabiodopsis Genome Initiative" (AGI) Locus Code.  Locus Codes have a predictable structure

* The first character is "A" or "a"
* The second character is "T" or "t"
* The third character is the chromosome number (between 1 and 5)
* The fourth character is "G" or "g"
* The remaining characters are a set of 5 digits (between 0 and 9)

**For Example:  At1G22934**

The final column in the dataset is the PubMed ID for the published paper describing this phenotype.  PubMed IDs are all digits (0-9) but the length of the ID is unpredictable.

# Exam Problems - TIME 2 hours
**You may use ANY RESOURCE YOU WANT - including Google, or the course notes.  **
**Please DO NOT discuss your answers with each other until after the exam is complete.**

please create a Notebook "Text box" describing your answer, and a "code box" containing the command, for problems #2 to #10, below (not #1 or #11)


## Problem 1

* Copy this data file to your copy of the Exams Git ( https://github.com/YourUsername/WilkinsonExams )
* Create a new Jupyter Notebook in that folder called "Exam_1_Answers" using the 'bash' kernel 
* Do the rest of this exam inside of that the "Exam_1_Answers" notebook

## Problem 2

Create a directory listing command that shows:
* the ownership of the file
* it's file-size, in Megabytes (i.e. human readable)
* in words (in the text box), describe the _permissions_ on that file (read/write/execute) for users, groups, and "anyone"

In [1]:
ls -lh Locus_Germplasm_Phenotype_20130122.txt

-rw-rw-r-- 1 osboxes osboxes 1.2M Sep  5 11:31 Locus_Germplasm_Phenotype_20130122.txt


**Answer: The file Exam_1_Answers.ipynb has the following _permissions_:**
* For users: read and write.
* For groups: read and write.
* For anyone: read.

I used -lh because -l already list all the parameters needed for the answer (ownership and file-size) and I have just added the argument -h to show it in Megabytes (makig it human readable).

## Problem 3

Create a command that outputs only the "header" line

In [2]:
head -1 Locus_Germplasm_Phenotype_20130122.txt

Locus_name)	Germplasm_name	phenotype	pubmed_id


The header, which describes the type of data of each column, is in the first line of the file in this file. For that reason,  I used the command head and specify -1, which implies that only the first line of the archieve is shown.

## Problem 4

Create a command that outputs the total number of lines in the file

In [3]:
cat Locus_Germplasm_Phenotype_20130122.txt | wc -l

7216


As It is specify that the command outputs only the total number of lines in the file, I used pipes to display just the number.

I used "cat Locus_Germplasm_Phenotype_20130122.txt" to display the entire document and then "wc -l" to count all the lines of the document. 

It is also possible to use "wc -l Locus_Germplasm_Phenotype_20130122.txt" but it gives you the number of lines and the name of the file.

## Problem 5

* Create a command that writes ONLY the data lines (i.e. excludes the header!) to a new file called "Data_Only.csv"
* prove that your output file has the expected number of lines

In [4]:
cat Locus_Germplasm_Phenotype_20130122.txt | tail -7215 > Data_Only.csv
wc -l Data_Only.csv
#head -1 Data_Only.csv
#tail -1 Data_Only.csv
#head -2 Locus_Germplasm_Phenotype_20130122.txt
#tail -1 Locus_Germplasm_Phenotype_20130122.txt

7215 Data_Only.csv


To create the "Data_Only.csv" I used the following pipe: "cat Locus_Germplasm_Phenotype_20130122.txt | tail -7215 > Data_Only.csv" wich:
        1. prints the whole document of Locus_Germoplasm_(...).txt then pipe to 
        2. wich selects starting from the last line the lines that you specified in -number. As I know the document have 7216 lines from the last problem, selecting all minus one will give me all the data except the header and then pipe it into...
        3. the output file Data_only.csv with > and the name of the file.
        
Then, I output the number of lines of the Data_Only.csv with "wc -l Data_Only.csv" which tell you the lines of the document as said before. 

Finally, I checked that everything is correct and working properly with head and tail of the Data_Only.csv and Locus_Germplasm_Phenotype_20130122.txt. If both entries are the same, it means it worked correctly. I commented this final lines as they are not relevant to the problem it-self.

## Problem 6

* Create a command that shows all of the lines that have a phenotype including the word "root"

In [5]:
grep "root" -Ein Locus_Germplasm_Phenotype_20130122.txt

23:AT1G01550	bps1-2	Growth of bps1-2 mutants on CPTA-supplemented medium resulted in partial rescue of both leaf and root defects. bps1-2 mutants grown on control growth medium have small radialized leaves with very little vascular tissue, and very short misshapen knotted-looking roots. By contrast, bps1-2 mutants grown on CPTA-supplemented medium produced larger flattened leaves that contained primary and secondary veins, and smooth elongated roots.	17217459
24:AT1G01550	bps1-2	Mutant roots were abnormal; primary and lateral roots were short, root hairs formed close to the root apex, and root defects were most severe when plants were grown at low temperature.	15458645
26:AT1G01550	bps1-2	Under high light conditions (approximately 200 &mu;E.m-2.sec-1), CPTA-treated seedlings were completely photobleached and mutants showed rescue of both leaf and root development. Under moderate (approximately 100 &mu;E.m-2.sec-1) and low (30 &mu;E.m-2.sec-1) light intensities, seedlings are incomplete

251:AT1G08030	SALK_009847	Seedlings of tpst-1 had stunted roots and smaller cotyledons compared with the WT.  tpst-1 has an abnormally shaped root apical meristem due to disorganized cell division and expansion. At 15 days after germination, the leaves of the tpst-1 plants are small and pale green, the number of higher order veins is reduced, and the secondary veins often do not close.  At the flowering stage, tpst-1 has smaller rosettes, tiny leaves showing early senescence, shorter inflorescence, and a reduced number of flowers and siliques. The fertility and seed set of tpst-1 are normal.	19666544
253:AT1G08260	tilted1-4	Twenty-five percent of embryos show abnormal cell divisions during embryogenesis-large cells- decreased rate of cell division. Division of the hypohyseal cell abnormal leading to abnormal placement of the root pole. Homozygotes show delayed flowering, abnormal floraly phyllotaxy and abnormal ovules.	16278345
255:AT1G08370	dcp1-2	The homozygous progeny is seedling le

545:AT1G18350	bud1	bud1 plants develop significantly fewer  lateral roots, simpler venation patterns, and a quicker and greater curvature in the gravitropism assay. In addition, the bud1  plants have shorter hypocotyls at high temperature (29C) under light, which is a characteristic feature of defective auxin  action. Determination of tritium-labeled indole-3-acetic acid transport showed that the increased expression of MKK7 in bud1  or the repressed expression in MKK7 antisense transgenic plants causes deficiency or enhancement in auxin transport,  indicating that MKK7 negatively regulates polar auxin transport.	16377756
553:AT1G18800	nrp1-1 nrp2-1	Double mutant plants showed a short-root phenotype: roots grew like the wild-type roots until 6 d after germination (DAG). However, after 7 DAG, the elongation of the double mutant roots was dramatically reduced compared with the wild-type roots. The aerial organs (leaves, rosettes, inflorescences, flowers, fruits, and embryos) developed no

I don't know how to limit the "grep" search to the phenotype column exclusively.

But, using "grep "root" -in Locus_Germplasm_Phenotype_20130122.txt" you obtain the same anwser as no other column contains the word root. 

grep searches for the expression inside "" in a file. The arguments -i means that the search is case insensitive to record all the times root is in the beginning of a phrase i.e. and the -n specify to show the number of the line in the file.

## Problem 7

* Create a command that writes the AGI Locus Code for every line that has a phenotype including the word "root" to a file called: <br/>"Root-associated-Loci.txt"

In [6]:
grep "root" -i Locus_Germplasm_Phenotype_20130122.txt | grep -Eino "AT[0-9]G[0-9]{5}" > Root-associated-Loci.txt
wc -l Root-associated-Loci.txt
grep "root" -i Locus_Germplasm_Phenotype_20130122.txt > check.txt
wc -l check.txt

55 Root-associated-Loci.txt
56 check.txt


I used the following pipe:


The first column is the "Arabiodopsis Genome Initiative" (AGI) Locus Code.  Locus Codes have a predictable structure

* The first character is "A" or "a" -> A
* The second character is "T" or "t" -> T
* The third character is the chromosome number (between 1 and 5) -> [0-9]
* The fourth character is "G" or "g" -> G
* The remaining characters are a set of 5 digits (between 0 and 9) -> [0-9]{5}

## Problem 8

* Create a command that writes the PubMed ID for every line that has a phenotype including the word "root" to a file called: <br/>"Root-associated-Publications.txt"

In [7]:
grep "root" -i Locus_Germplasm_Phenotype_20130122.txt | grep -E "\s[0-9]+" > Root-associated-Publications.txt
grep "root" -i Locus_Germplasm_Phenotype_20130122.txt > check2.txt
wc -l Root-associated-Publications.txt
wc -l check2.txt


54 Root-associated-Publications.txt
56 check2.txt


The final column in the dataset is the PubMed ID for the published paper describing this phenotype.  PubMed IDs are all digits (0-9) but the length of the ID is unpredictable.

## Problem 9

* _**Control experiment**_:  You would hypothesize that genes associated with roots **should be found on all chromosomes.**  Find a way (one or more commands) to test this hypothesis.  In this dataset, is the hypothesis true? 

In [8]:
grep "root" -i Locus_Germplasm_Phenotype_20130122.txt | grep -Eino "AT[0-9]G[0-9]{5}" > Root-associated-Loci.txt
grep "AT1\w+" -Eino Root-associated-Loci.txt > Chr1.txt
grep "AT2\w+" -Eino Root-associated-Loci.txt > Chr2.txt
grep "AT3\w+" -Eino Root-associated-Loci.txt > Chr3.txt
grep "AT4\w+" -Eino Root-associated-Loci.txt > Chr4.txt
grep "AT5\w+" -Eino Root-associated-Loci.txt > Chr5.txt
wc -l Chr1.txt
wc -l Chr2.txt
wc -l Chr3.txt
wc -l Chr4.txt
wc -l Chr5.txt

55 Chr1.txt
0 Chr2.txt
0 Chr3.txt
0 Chr4.txt
0 Chr5.txt


The hypothesis that genes that are associated with root phenotypes are found in all chromosomes is false, using this code. 

I tested it using the file that we already used in exercise 7 with all the root associated Loci. Then I grep in this file for the specific chromosome using the ATn, where the n correspond to one of the five chromosomes of A. thaliana, and output it in a file name "Chrn.txt", where the n correspond to the chromosome number. Finally, I printed the lines that are in each file to see the distribution for each chromosome using the .txt corresponding to each chromosome. All the genes are stored in the chromosome 1.

## Problem 10

* This dataset is from the most popular Arabidopsis database on earth!  **I did not change it!**  What does your answer to Problem 9 tell you?  Can you explain it?  What do you learn from this control that you should remember for the rest of your life?!

I know it is incorrect, because searching with Cntr + F in the original .txt shows 1501 matches and I only see 55 genes. I don't know what's wrong, but it shows the importance of good controls.

## Problem 11

* push your answers to GitHub and give me your GitHub username before you leave the class. I will clone your repositories and grade your answers.
